### PR TITLE
Implemented const and reverse iterator, insert.

### PR DESCRIPTION
**Pull request by mbiswal94**
Has a small error related to iterator

### CODE REVIEW

File: tests/vector_test.h [hunk 1]
```diff
@@ -1,8 +1,10 @@
-#include "buffer.h"
+#include "../buffer.h"
```
Comment (rajathshashidhara) [inline tests/vector_test.h:1]: not a good idea. we should do something better..

Comment (devenbansod) [inline tests/vector_test.h:1]: May be a `./includes` directory ?

File: tests/vector_test.h [hunk 2]
```diff
@@ -1,8 +1,10 @@
...
 
 template <typename T>
 class vector
 {
+public:
```
Comment (rajathshashidhara) [inline tests/vector_test.h:6]: why is this required ?

Comment (mbiswal94) [inline tests/vector_test.h:6]: size_type forward declaration is required to make sz of type: size_type

File: tests/vector_test.h (hunk 1 above)
Comment (rajathshashidhara) [inline tests/vector_test.h:1]: this will be handled in the make file using the -I option.

File: tests/vector_test.h (hunk 2 above)
Comment (rajathshashidhara) [inline tests/vector_test.h:6]: oh! yes.

File: tests/vector_test.h (hunk 1 above)
Comment (rajathshashidhara) [inline tests/vector_test.h:1]: i will merge it if you change this line.

File: tests/vector_test.h [hunk 3]
```diff
@@ -188,11 +344,54 @@ void vector<T>::insert(vector<T>::iterator position, const T& val) {
...
+		memcpy(temp_ptr, temp_buff + block_size - element_size, element_size);
+
+		void *new_buff = malloc(block_size);
+		memcpy(new_buff,temp_buff + block_offset, block_size - element_size);
+		memcpy(temp_buff + block_offset + element_size, new_buff, block_size - element_size);
+		memcpy(temp_buff + block_offset, &val, element_size);
+		buffered_file.writeBlock(block_number,buff);
+
+		position.index = block_number*num_elements_per_block;
+		insert(position, tempT);
```
Comment (rajathshashidhara) [inline tests/vector_test.h:395]: I will prefer the iterative version.
